tsnet/tsnet_test.go

```diff
@@ -344,3 +353,83 @@ func TestTailscaleIPs(t *testing.T) {
 			sIp4, upIp4, sIp6, upIp6)
 	}
 }
+
+func TestFunnel(t *testing.T) {
+	ctx, dialCancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer dialCancel()
+
+	controlURL := startControl(t)
+	s1, _ := startServer(t, ctx, controlURL, "s1")
+	s2, _ := startServer(t, ctx, controlURL, "s2")
+
+	s1lc := must.Get(s1.LocalClient())
+	s1st := must.Get(s1lc.StatusWithoutPeers(ctx))
+	peer6 := s1st.Self.PeerAPIURL[1] // IPv6
+	t.Logf("using peerAPI %q", peer6)
+
+	ln := must.Get(s1.Listen("tcp", ":8080"))
+	defer ln.Close()
+	wantSrcAddrPort := netip.MustParseAddrPort("127.0.0.1:1234")
+	wantTarget := ipn.HostPort("s1:8080")
+	go func() {
+		srv := &http.Server{
+			ConnContext: func(ctx context.Context, c net.Conn) context.Context {
+				if fc, ok := c.(*ipn.FunnelConn); !ok {
+					t.Errorf("ConnContext called with non-FunnelConn")
+				} else if fc.Src != wantSrcAddrPort {
+					t.Errorf("ConnContext called with wrong SrcAddrPort; got %v, want %v", fc.Src, wantSrcAddrPort)
+				} else if fc.Target != wantTarget {
+					t.Errorf("ConnContext called with wrong Target; got %q, want %q", fc.Target, wantTarget)
+				}
+				return ctx
+			},
+			Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				fmt.Fprintf(w, "hello")
+			}),
+		}
+		srv.Serve(ln)
+	}()
+
+	// Enable Funnel for s1:8080.
+	must.Do(s1lc.SetServeConfig(ctx, &ipn.ServeConfig{
+		AllowFunnel: map[ipn.HostPort]bool{
+			wantTarget: true,
+		},
+	}))
```

Comment:
Test negative case before setting serve config?

Reply:
Not applicable anymore as I'm using `ListenFunnel`